[docs] EAS Build - managed version support (#18440)
expo/expo: docs/pages/build-reference/app-versions.md

## Examples

<Collapsible summary="Remote version source">

With this **eas.json**, the version for all builds will be based on the value stored on EAS servers, and the version will be incremented remotely and only when building with `production` profile.

```json
{
  "cli": {
    "appVersionSource": "remote"
  },
  "build": {
    "staging": {
      "distribution": "internal",
      "android": {
        "buildType": "apk"
      }
    },
    "production": {
      "autoIncrement": true
    }
  }
}
```

</Collapsible>

<Collapsible summary="Local version source">

With this **eas.json**, the version for all builds will be based on the value from **app.json** or native code. When you build using `production` profile, the version will be incremented in the local code before the build.

```json
{
  "cli": {
    "appVersionSource": "local"
  },
  "build": {
    "staging": {
      "distribution": "internal",
      "android": {
        "buildType": "apk"
      }
    },
    "production": {
      "autoIncrement": true
    }
  }
}
```

</Collapsible>

## Remote version source

> You can configure your project to rely on EAS servers to store and manage the version of your app, to do that add `{ "cli": { "appVersionSource": "remote" } }` in your **eas.json**. The remote version will be initialized with the value from a local project — if you would like to explicitly set the value directly, or EAS CLI is not able to detect what version an app is on, you can use the `eas build:version:set` command. EAS stores version information scoped by account, slug, platform, and application ID/bundle identifier — so, for example, if you are building variants with different application IDs or bundle identifiers, versioning will be independent for each of them.

Reviewer: ```suggestion
You can configure your project to rely on EAS servers to store and manage the version of your app. Add `{ "cli": { "appVersionSource": "remote" } }` in your **eas.json**. The remote version will be initialized with the value from the local project. If you would like to explicitly set the value directly, or EAS CLI is not able to detect what version the app is on, you can use the `eas build:version:set` command. EAS stores version information scoped by account, slug, platform, and application ID/bundle identifier — so, for example, if you are building variants with different application IDs or bundle identifiers, versioning will be independent for each of them.
```

When using a remote app version source, the values in **app.json** will not be updated when the version is incremented remotely, and so the local and remote values will fall out of sync. The remote source values will be set on the native project when running a build, and they are the source of truth  — however, the values specified in your **app.json** will be present in `Constants.expoConfig` and `Constants.manifest` exposed by `expo-constants`. Use `expo-application` to determine your application version at runtime instead, and remove `versionCode`/`buildNumber` from your **app.json**.

### Limitations

- `eas build:version:sync` command on Android does not support bare projects with multiple flavors, but the rest of the remote versioning functionality should work with all projects.
- `autoIncrement` does not support the `version` option.
- It's not supported if you are using EAS Update and runtime policy set to `"runtimeVersion": { "policy": "nativeVersion" }`. For similar behavior, use the `"appVersion"` policy instead.

### Recommended workflow

> The main goal of this feature is to avoid manual changes to the project every time you are uploading a new archive to run it on TestFlight / Play Store testing channels. When you are doing a production release, the user-facing version change should be explicit.

Reviewer 2: ```suggestion
The main goal of this feature is to avoid manual changes to the project every time you are uploading a new archive to run it on TestFlight/Play Store testing channels. When you are doing a production release, the user-facing version change should be explicit.
```

Let's refrain ourselves from adding extra spacing before and after a "/".  😄

We recommend updating `version` field after a new build goes live in the store, especially if you are using `expo-updates` with an automatic runtime version policy. This marks the beginning of a new development cycle for a new version of your app. [Learn more about deployment patterns](/eas-update/deployment-patterns.md).

## Local version source

> By default, the source of truth for a project versions is the local project source code itself. In this case, EAS does not write to the project, it reads the values and builds projects as they are.

Reviewer: ```suggestion
The source of truth for project versions is the local project's source code. In this case, EAS does not write to the project, it reads the values and builds projects as they are.
```

Author: ```suggestion
By default, the source of truth for project versions is the local project source code itself. In this case, EAS does not write to the project, it reads the values and builds projects as they are.
```
"By default" is essential here, both local and remote values can be a source of truth depending on eas.json config

You may opt in to auto incrementing versions locally with the `autoIncrement` option on a build profile, but it comes with some limitations.

> In the case of bare React Native projects, values in native code take priority, and `expo-constants` and `expo-updates` read values from **app.json**. If you rely on version values from a manifest, you should keep them in sync with native code. Keeping these values in sync is especially important if you are using EAS Update with the runtime policy set to `"runtimeVersion": { "policy": "nativeVersion" }`, because it mismatched versions may result in the delivery of updates to the wrong version of an application. We recommend using `expo-application` to read the version instead of depending on values from **app.json**.

Reviewer: ```suggestion
In the case of bare React Native projects, values in native code take precedence, and `expo-constants` and `expo-updates` read values from **app.json**. If you rely on version values from a manifest, you should keep them in sync with native code. Keeping these values in sync is especially important if you are using EAS Update with the runtime policy set to `"runtimeVersion": { "policy": "nativeVersion" }`, because mismatched versions may result in the delivery of updates to the wrong version of an application. We recommend using `expo-application` to read the version instead of depending on values from **app.json**.
```

If you want to build your project in Android Studio or Xcode, you can update your local project with the remote versions using `eas build:version:sync`.

Enabling `autoIncrement` option in this mode is only possible for `versionCode`/`buildNumber`, make sure to remove those values from **app.json** otherwise, they will still show up in the manifest exposed by `expo-constants` even though the app itself will be on a correct version.

### Limitations

- `eas build:version:sync` command on Android does not support bare projects with multiple flavors, but the rest of the remote versioning functionality should work with all projects.
- `autoIncrement` does not support the `version` option.
- It's not supported if you are using EAS Update and runtime policy set to `"runtimeVersion": { "policy": "nativeVersion" }`.

### Recommended workflow

> The main goal of this feature is to avoid manual changes to the project every time you are uploading a new archive to run it on a TestFlight or Play Store testing channels, but when you are doing a production release, the change should be explicit. We recommend always updating `version` field after a new build goes live in the store, especially if you are using `expo-updates` with an automatic runtime version policy.

Reviewer 3: ```suggestion
The main goal of this feature is to avoid manual changes to the project every time you are uploading a new archive to run it on TestFlight / Play Store testing channels. When you are doing a production release, the user-facing version change should be explicit.

We recommend updating `version` field after a new build goes live in the store, especially if you are using `expo-updates` with an automatic runtime version policy. This marks the beginning of a new development cycle for a new version of your app. [Learn more about deployment patterns](/eas-update/deployment-patterns.md).
```

## Local version source

> By default all projects keep versions locally, EAS itself does not interact with it and just builds projects as they are, unless `autoIncrement` option is enabled. In the case of bare React Native projects, values in native code take priority, but `expo-constants` and `expo-updates` read values from **app.json**. If you rely on version values from a manifest, you should keep them in sync with native code. It's especially important if you are using EAS Update with runtime policy set to `"runtimeVersion": { "policy": "nativeVersion" }`, because it might result in the delivery of updates to the wrong version of an application.

Reviewer 3: ```suggestion
By default, the source of truth for a project versions is the local project source code itself. In this case, EAS does not write to the project, it reads the values and builds projects as they are.

You may opt in to auto incrementing versions locally with the `autoIncrement` option on a build profile, but it comes with some limitations.

In the case of bare React Native projects, values in native code take priority, and `expo-constants` and `expo-updates` read values from **app.json**. If you rely on version values from a manifest, you should keep them in sync with native code. Keeping these values in sync is especially important if you are using EAS Update with the runtime policy set to `"runtimeVersion": { "policy": "nativeVersion" }`, because it mismatched versions may result in the delivery of updates to the wrong version of an application. We recommend using `expo-application` to read the version instead of depending on values from **app.json**.
```

### Limitations

> - With `autoIncrement` option, you need to commit your changes on every build, if you want the version change to persist. It's especially problematic when building from CI.

Reviewer 3: ```suggestion
- With `autoIncrement`, you need to commit your changes on every build if you want the version change to persist. This can be difficult to coordinate when building on CI.
```

- `autoIncrement` is not supported if you are using a dynamic config (**app.config.js**).

> - For bare React Native projects with gradle configuration that supports multiple flavors, EAS CLI is not able to read or modify the version, so `autoIncrement` option is not supported and versions will not be listed on the build page.

Reviewer 3: ```suggestion
- For bare React Native projects with Gradle configuration that supports multiple flavors, EAS CLI is not able to read or modify the version, so `autoIncrement` option is not supported and versions will not be listed in the build details page on [expo.dev](https://expo.dev).
```
